### PR TITLE
Improve normalize_chunks calculation for "auto" setting

### LINKED ISSUE #11341: Improve how normalize_chunks selects chunk sizes if auto is given
Specifying auto for a dimension when new chunks are selected like in rechunking allows us to choose a chunk size that we think is a good fit and efficient. 

This is for now an issue to collect thoughts from others and feedback, I am pretty sure that I haven't found all suboptimal corner cases yet.

Assume we have the following array:

```
import dask.array as da

arr = da.random.random((5000, 300, 372), chunks=(600, 150, 172))
```

Triggering a Rechunk operation to -1 and auto for dimensions 2 and 3 results in the following chunks:

```
x = arr.rechunk((-1, "auto", "auto"))
```

Screenshot 2024-08-20 at 15 34 19

```normalize_chunks`` targets the 128MiB that is the default for auto pretty aggressively and creates a significantly more complex task graph than necessary if we would compromise a little bit on the in memory size of the chunks.

Choosing 50 and (58, 57, 57) per input chunk would allows us to split every chunk into 3 pieces along every axis and thus avoid including the neighbouring chunks alongside this axis.

I am pretty sure that there are better examples where this makes things very complex, but this should illustrate the general issue.

The proposal is to compromise a little bit on the chunk size side of things to create a simpler topology if possible.

### PR DESCRIPTION
**Pull request by phofl**
- [ ] Closes #11341
- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

### CODE REVIEW

File: dask/array/core.py [hunk 1]
```diff
@@ -3228,54 +3230,104 @@ def auto_chunks(chunks, shape, limit, dtype, previous_chunks=None):
...
+                        dimension_result.extend(new_chunks)
+
+                        if max(dimension_result) == 1:
+                            # See if we can split up other axes further
+                            multiplier_remaining = _trivial_aggregate(a)
+
                 else:
-                    result[a] = round_to(proposed, ideal_shape[a])
+                    dimension_result, new_chunk = [], 0
+                    for c in previous_chunks[a]:
```
Comment (hendrikmakait) [inline dask/array/core.py:3299]: Do we need to check this within the for-loop?

File: dask/array/core.py [hunk 2]
```diff
@@ -3228,54 +3230,104 @@ def auto_chunks(chunks, shape, limit, dtype, previous_chunks=None):
...
+                        dimension_result.append(new_chunk)
+
+                result[a] = tuple(dimension_result)
 
             # recompute how much multiplier we have left, repeat
-            multiplier = _compute_multiplier(limit, dtype, largest_block, result)
+            if multiplier_remaining:
+                multiplier = _compute_multiplier(
+                    limit, dtype, largest_block, median_chunks
+                )
```
Comment (hendrikmakait) [inline dask/array/core.py:3328]: nit: I've been thinking about the directional `left` here
```suggestion
            if multiplier_remaining:
```

File: dask/array/core.py [hunk 3]
```diff
@@ -3297,6 +3349,23 @@ def auto_chunks(chunks, shape, limit, dtype, previous_chunks=None):
         return tuple(chunks)
 
 
+def _split_up_single_chunk(c, this_chunksize_tolerance, max_chunk_size, proposed):
+    # Calculate by what factor we have to split this chunk
+    m = c / proposed
+    if math.ceil(m) / m > this_chunksize_tolerance:
+        # We want to smooth things potentially if rounding up would change the result
```
Comment (hendrikmakait) [inline dask/array/core.py:3356]: nit: Would it be possible to add type hints?

File: dask/array/core.py (hunk 1 above)
Comment (phofl) [reply dask/array/core.py:3299]: Can you say more?

File: dask/array/core.py (hunk 2 above)
Comment (phofl) [reply dask/array/core.py:3328]: done that

File: dask/array/core.py (hunk 1 above)
Comment (hendrikmakait) [reply dask/array/core.py:3299]: It looks like VS Code messed up the line this is attached to.

File: dask/array/core.py [hunk 4]
```diff
@@ -3228,54 +3230,104 @@ def auto_chunks(chunks, shape, limit, dtype, previous_chunks=None):
...
+                                c, this_chunksize_tolerance, max_chunk_size, proposed
+                            )
+
+                        cache[c] = new_chunks
+                        dimension_result.extend(new_chunks)
+
+                        if max(dimension_result) == 1:
+                            # See if we can split up other axes further
+                            multiplier_remaining = _trivial_aggregate(a)
+
```
Comment (hendrikmakait) [inline dask/array/core.py:3296]: Do we need to perform this check within the for-loop? From what I understand, we want to understand at the end of the loop whether dimension_result is all 1s.

Comment (phofl) [reply dask/array/core.py:3296]: That was me being stupid and missing an intendation, so no we can do this outside

File: dask/array/core.py (hunk 3 above)
Comment (phofl) [reply dask/array/core.py:3358]: done

**Comment by rjzamora** [comment]
Rerun tests

**Review by hendrikmakait — APPROVED** [review]
Thanks, @phofl! The changes look good, this just needs some conflicts resolved.

**Pull request merged**